### PR TITLE
[WIP] Nightshade + Network

### PR DESCRIPTION
**Pull request by mfornet**
Running nightshade using real network. Currently, `alphanet` just launch an instance of nightshade to run consensus on a single block (which is not posted to anywhere). Starting information such as AccountsId and PublicKeys are generated beforehand.

**Current status**: Consensus is reached but nodes never stop the task because the stop signal is never sent. We should send this signal after committing to some outcome and see that more than 2/3 of authorities have committed too, earlier exiting may cause that remaining authorities don't reach consensus.

This branch depends on `nightshadev2`. **NOTE** this merge is into `nigthsadev2` to separate commits that belongs here only. This should be merged into `master` after `nightshadev2` is merged and this works fine.

### CODE REVIEW

File: node/network/src/nightshade_protocol.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+fn forward_msg<T>(ch: Sender<T>, el: T)
+    where T: Send + 'static,
+{
+    let task =
+        ch.send(el).map(|_| ()).map_err(|e| warn!(target: "network", "Error forwarding {}", e));
+    tokio::spawn(task);
+}
+
+// TODO: Document this function
+pub fn start_peer(
```
Comment (nearmax) [inline node/network/src/nightshade_protocol.rs]: This function breaks the encapsulation of the low-level network details. It creates PeerManager, passes it to the code that starts alphanet (breaking the encapsulation of the PeerManager), alphanet waits until there are enough peers, then passes PeerManager back to the network layer and starts the consensus.

Instead it should follow the same setup as existing testnet and devnet where initialization of coroutines happens without waiting for special events, like sufficient number of connected peers.

File: node/network/src/nightshade_protocol.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+/// Spawn network task that process incoming and outgoing gossips for nightshade consensus
+///
+/// Args:
+pub fn spawn_consensus_network(
+    pm: PeerManager,
+    inc_msg_rx: Receiver<(CryptoHash, Vec<u8>)>,
+    inc_gossip_tx: Sender<Gossip<ChainPayload>>,
+    out_gossip_rx: Receiver<Gossip<ChainPayload>>,
+    authority_map: HashMap<AuthorityId, AccountId>,
+) {
```
Comment (nearmax) [inline node/network/src/nightshade_protocol.rs]: `PeerManager` should not be passed from outside, but should be created inside.

File: node/alphanet/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    if let Ok(lvl) = env::var("RUST_LOG") {
+        builder.parse(&lvl);
+    }
+    if let Err(e) = builder.try_init() {
+        warn!(target: "client", "Failed to reinitialize the log level {}", e);
+    }
+}
+
+// TODO: Explain this function and their arguments
+fn run_node(authority: usize,
```
Comment (nearmax) [inline node/alphanet/src/lib.rs]: Also, for consistency, I suggest following the existing naming, e.g. name this method `start_from_configs`.

File: node/alphanet/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        let payload: ChainPayload = ChainPayload {
+            transactions: vec![],
+            receipts: vec![],
+        };
+
+        let (inc_gossip_tx, inc_gossip_rx) = mpsc::channel(1024);
+        let (out_gossip_tx, out_gossip_rx) = mpsc::channel(1024);
+        let (consensus_tx, _consensus_rx) = mpsc::channel(1024);
+
+        let control_tx = spawn_nightshade_task(
```
Comment (nearmax) [inline node/alphanet/src/lib.rs]: Let us adopt the following convention:
If we have two components A and B that are both owned by component C, such that A and B are connected using channel (Receiver, Sender), then we construct (Receiver, Sender) in C and pass the corresponding ends to A and B.

**Comment by nearmax** [comment]
I have pushed some changes, but it is still WIP. Things that need to be done:
* `run_node` should accept configs only;
* `PeerManager` should not be exposed outside the network;

Please follow testnet crate https://github.com/nearprotocol/nearcore/blob/e5ceb0269b448fb5a962bc33af10df6bc4d800d9/node/testnet/src/lib.rs and testnet integration test https://github.com/nearprotocol/nearcore/blob/e5ceb0269b448fb5a962bc33af10df6bc4d800d9/tests/test_testnet.rs for alphanet.

**Comment by mfornet** [comment]
@nearmax If we are using the same NightshadeTask instance to run several consensus two things arise:
* The preferred proposal is from the participant with the highest `AuthorityId`. Here we can use other "un-biasable" tie-breaker (so no rotation is needed) or rotate participants frequently.  If validators rotate then `public_keys` (bls and non-bls) must be updated.
* It makes sense that consensus on different blocks are running at the same time concurrently. If a participant has already committed to some value should not abandon consensus until at least 2/3 other participants arrive to consensus too, but she can start working on the next block immediately.

**Comment by nearmax** [comment]
> * The preferred proposal is from the participant with the highest `AuthorityId`. Here we can use other "un-biasable" tie-breaker (so no rotation is needed) or rotate participants frequently.  If validators rotate then `public_keys` (bls and non-bls) must be updated.

My understanding was that we need to do two things anyway:
* Rotate participants, meaning have a new set of participants for each block, where participants for block X and block X+1 have large, e.g. 60-95% overlap;
* Reshuffle the ranking of participants on each block;

So currently we need to send all initialization parameters (those that we pass to NS constructor) with Reset signal.

> * It makes sense that consensus on different blocks are running at the same time concurrently. If a participant has already committed to some value should not abandon consensus until at least 2/3 other participants arrive to consensus too, but she can start working on the next block immediately.

I totally like the idea of having NS overlap. On higher level, having consensuses overlap sounds desirable, but technically I think (and judging from all different ways I tried implementing TxFlow task in the past) it is going to be extremely difficult to implement without errors. The more things are happening in parallel the more corner cases can happen. I suggest we do the simpler implementation first (where we have only one instance of NS running at the given time) and do the complex implementation later.

**Comment by nearmax** [comment]
Merging intermediate code with master.

**Pull request merged**